### PR TITLE
Fix several stretch mode issues

### PR DESCRIPTION
**Pull request by gregsfortytwo**
This PR resolves several issues others have encountered while testing the new stretch mode.

### CODE REVIEW

**Comment by gregsfortytwo** [comment]
I have a few remaining issues I've identified (with pool creation and multiple data center failures) I'm still working on, but want to get these looked at.

**Review by jdurgin — APPROVED** [review]
thanks for the early update, the fixes so far look good. just one comment - the downstream bug references don't belong in the upstream commits

**Comment by gregsfortytwo** [comment]
> thanks for the early update, the fixes so far look good. just one comment - the downstream bug references don't belong in the upstream commits

Hmm, should I copy them into our redmine then? They're public bugs and that rhbz is the original source for the issue on unreleased code so it seemed right to me.

**Comment by gregsfortytwo** [comment]
@jdurgin ^ forgot to tag

**Comment by jdurgin** [comment]
if you think it needs to be tracked in a ticket, yes. I don't think it's necessary since these aren't being backported upstream though

File: src/mon/MonMap.cc [hunk 1]
```diff
@@ -41,7 +41,11 @@ void mon_info_t::encode(ceph::buffer::list& bl, uint64_t features) const
   uint8_t v = 5;
   uint8_t min_v = 1;
   if (!crush_loc.empty()) {
-    min_v = 5;
+    // we added crush_loc in version 5, but need to let old clients decode it
+    // so just leave the min_v at version 4. Monitors are protected
+    // from misunderstandings about location because setting it is blocked
+    // on FEATURE_PNGING
```
Comment (neha-ojha) [inline src/mon/MonMap.cc:47]:
```suggestion
    // on FEATURE_PINGING
```

File: src/mon/MonMap.cc [hunk 2]
```diff
@@ -369,6 +373,9 @@ void MonMap::print(ostream& out) const
   out << "min_mon_release " << to_integer<unsigned>(min_mon_release)
       << " (" << min_mon_release << ")\n";
   out << "election_strategy: " << strategy << "\n";
+  if (disallowed_leaders.size()) {
+    out << "disallowed_leaders" << disallowed_leaders << "\n";
```
Comment (neha-ojha) [inline src/mon/MonMap.cc:377]:
```suggestion
    out << "disallowed_leaders " << disallowed_leaders << "\n";
```

**Review by neha-ojha — APPROVED** [review]

**Comment by gregsfortytwo** [comment]
http://pulpito.front.sepia.ceph.com/gregf-2020-11-06_17:45:44-rados-wip-stretch-fixes-116-2-distro-basic-smithi/

8 failed: 5596825, 5596919, 5597112 are https://tracker.ceph.com/issues/47024; 5596861 is an infra issue "machine smithi186.front.sepia.ceph.com is locked by scheduled_teuthology@teuthology, not scheduled_gregf@teuthology"; 5596878 is https://tracker.ceph.com/issues/45327; 5597065 is https://tracker.ceph.com/issues/47719; 5597164 and 5597197 had a few quorums form without all monitors but they got together in the end. I see mon_pings happening but an election message was sent on one side and didn't show up on the other?

Of the dead jobs, 5596835's failure reason was "SSH connection to smithi186 was lost: 'sudo yum -y install ceph-radosgw'" and 5596839 had some kind of ansible failure spawn terrible messages to the log.

**Comment by alfonsomthd** [comment]
@gregsfortytwo Do you remember if Ceph API tests jenkins job passed?
It seems that due to this change:
https://github.com/ceph/ceph/pull/37915/files?file-filters%5B%5D=.cc&file-filters%5B%5D=.h&file-filters%5B%5D=.t#diff-bfd87f36386e8f1a3e7ca8f6f2b1a212e52d9cf188c0fa73193dd5eb751a54caR2682
Now API test is failing:
```
2020-11-09 15:53:32,068.068 INFO:__main__:======================================================================
2020-11-09 15:53:32,068.068 INFO:__main__:FAIL: test_full_health (tasks.mgr.dashboard.test_health.HealthTest)
2020-11-09 15:53:32,069.069 INFO:__main__:----------------------------------------------------------------------
2020-11-09 15:53:32,069.069 INFO:__main__:Traceback (most recent call last):
2020-11-09 15:53:32,069.069 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/mgr/dashboard/helper.py", line 454, in assertSchema
2020-11-09 15:53:32,069.069 INFO:__main__:    return _validate_json(data, schema)
2020-11-09 15:53:32,069.069 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/mgr/dashboard/helper.py", line 668, in _validate_json
2020-11-09 15:53:32,069.069 INFO:__main__:    for key, sub_schema in schema.sub_elems.items()
2020-11-09 15:53:32,069.069 INFO:__main__:tasks.mgr.dashboard.helper._ValError: In `input['mon_status']`: unknown keys: {'stretch_mode'}
2020-11-09 15:53:32,070.070 INFO:__main__:
2020-11-09 15:53:32,070.070 INFO:__main__:During handling of the above exception, another exception occurred:
2020-11-09 15:53:32,070.070 INFO:__main__:
2020-11-09 15:53:32,070.070 INFO:__main__:Traceback (most recent call last):
2020-11-09 15:53:32,070.070 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/mgr/dashboard/test_health.py", line 272, in test_full_health
2020-11-09 15:53:32,070.070 INFO:__main__:    self.assertSchema(data, schema)
2020-11-09 15:53:32,070.070 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/mgr/dashboard/helper.py", line 456, in assertSchema
2020-11-09 15:53:32,070.070 INFO:__main__:    self.assertEqual(data, str(e))
2020-11-09 15:53:32,070.070 INFO:__main__:AssertionError: {'health': {'status': 'HEALTH_OK', 'check[310536 chars]: 0}} != "In `input['mon_status']`: unknown keys: {'stretch_mode'}"
```

**Comment by neha-ojha** [comment]
@gregsfortytwo https://github.com/ceph/ceph/pull/37997 is a follow-on fix for https://tracker.ceph.com/issues/48155

**Comment by gregsfortytwo** [comment]
@alfonsomthd hmm the api tests failed but when I investigated on an earlier version of this I got the impression it was some rook installation error, not Ceph at all...not sure.

**Comment by jdurgin** [comment]
you can see the run under 'view details' when it merged above: https://jenkins.ceph.com/job/ceph-api/6296/consoleFull#-10422008793f609609-74b6-4c11-b0c1-dab0624390b5

If you look at the 'failed test', 'indication 4', you'll see:

2020-11-06 07:03:13,747.747 INFO:__main__:tasks.mgr.dashboard.helper._ValError: In `input['mon_status']`: unknown keys: {'stretch_mode'}

This was causing all the api test runs to fail yesterday until Alfonso's fix merged

**Pull request merged**